lib/rules/no-dupe-disjunctions.ts

```diff
@@ -789,17 +791,22 @@ export default createRule("no-dupe-disjunctions", {
                         break
 
                     case "Overlap":
-                        context.report({
-                            node,
-                            loc: getRegexpLocation(result.alternative),
-                            messageId: "overlap",
-                            data: {
-                                exp,
-                                cap,
-                                others,
-                                expr: faToSource(result.overlap, flags),
-                            },
-                        })
+                        if (
+                            isStared(result.alternative) ||
+                            getUsageOfPattern() !== UsageOfPattern.partial
```

Comment:
This is the wrong place to check for partial patterns. The `reportResult` has the sole purpose of reporting.

Also, this is wrong. We detect partial patterns to report overlaps that might cause exp backtracking. Example:

```ts
const foo = /[ab]|[bc]/.source;
const bar = RegExp(`(${foo})+$`);
```

However, I'm not sure what the _correct_ behavior would be. The only thing I know for certain is:

- `report: "all"` should not be affected by partial patterns. `report: "all"` should just report everything it finds.

But there are still unanswered questions:

- Partial patterns and `PrefixSubset` results. E.g. ``var a = /a|aa/.source; var b = RegExp(`\\b${a}\\b`);``. In partial patterns, `PrefixSubset` results _might_ be false positives. We can't know for certain. Should we report them?

- _Potential_ cases of exp backtracking. We don't know whether overlap in partial patterns causes exponential backtracking. It can (see example above) but it also might not. We can't be certain.

  One solution might be to add another option like `reportPotentialExponentialBacktracking`. If set to `true`, all potential cases of exponential backtracking in partial patterns will be reported.

  Another solution is to remove the `alwaysReportExpoentialBacktracking: boolean` option (let's call this option `AREB` from now on...) and replace it with a new option `reportExpoentialBacktracking: "none" | "certain" | "potential"`. `none` is the current behavior of `AREB: false`, `certain` is the current behavior of `AREB: true`, and `potential` is the `AREB: true` + the potential cases of partial patterns.

Thoughts @ota-meshi?

Reply:
I like the idea of `reportExpoentialBacktracking`. I will implement the `reportExpoentialBacktracking` option.

> Partial patterns and PrefixSubset results. E.g. `var a = /a|aa/.source; var b = RegExp(`\\b${a}\\b`);`. In partial patterns, PrefixSubset results might be false positives. We can't know for certain. Should we report them?

I didn't notice it. We probably need a similar option, `reportPrefixSubset: "certain" | "potential"`.

I will fix a bug that doesn't check template literals.